executor/prepared.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"sort"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/executor/converter"
+	"github.com/pingcap/tidb/field"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/optimizer"
+	"github.com/pingcap/tidb/optimizer/evaluator"
+	"github.com/pingcap/tidb/optimizer/plan"
+	"github.com/pingcap/tidb/parser"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/stmt"
+)
+
+type paramMarkerSorter struct {
+	markers []*ast.ParamMarkerExpr
+}
+
+func (p *paramMarkerSorter) Len() int {
+	return len(p.markers)
+}
+
+func (p *paramMarkerSorter) Less(i, j int) bool {
+	return p.markers[i].Offset < p.markers[j].Offset
+}
+
+func (p *paramMarkerSorter) Swap(i, j int) {
+	p.markers[i], p.markers[j] = p.markers[j], p.markers[i]
+}
+
+type paramMarkerExtractor struct {
+	markers []*ast.ParamMarkerExpr
+}
+
+func (e *paramMarkerExtractor) Enter(in ast.Node) (ast.Node, bool) {
+	return in, false
+}
+
+func (e *paramMarkerExtractor) Leave(in ast.Node) (ast.Node, bool) {
+	if x, ok := in.(*ast.ParamMarkerExpr); ok {
+		e.markers = append(e.markers, x)
```

Comment:
The extractor will not be reused, and the param markers will be visited exactly once, so we don't need to clear it.